sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java

```diff
@@ -0,0 +1,232 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.google.common.collect.Iterables;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import org.apache.beam.sdk.coders.IterableCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.RunnableOnService;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.testing.TestStream;
+import org.apache.beam.sdk.testing.UsesStatefulParDo;
+import org.apache.beam.sdk.testing.UsesTestStream;
+import org.apache.beam.sdk.testing.UsesTimersInParDo;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Test Class for {@link BatchingParDo}. */
+@RunWith(JUnit4.class)
+public class BatchingParDoTest implements Serializable {
+  private static final int BATCH_SIZE = 5;
+  private static final long NUM_ELEMENTS = 10;
+  private static final int ALLOWED_LATENESS = 0;
+  private static final Logger LOGGER = LoggerFactory.getLogger(BatchingParDoTest.class);
+  @Rule public transient TestPipeline pipeline = TestPipeline.create();
+  private transient ArrayList<KV<String, String>> data = createTestData();
+
+  private static ArrayList<KV<String, String>> createTestData() {
+    String[] scientists = {
+      "Einstein",
+      "Darwin",
+      "Copernicus",
+      "Pasteur",
+      "Curie",
+      "Faraday",
+      "Newton",
+      "Bohr",
+      "Galilei",
+      "Maxwell"
+    };
+    ArrayList<KV<String, String>> data = new ArrayList<>();
+    for (int i = 0; i < NUM_ELEMENTS; i++) {
+      int index = i % scientists.length;
+      KV<String, String> element = KV.of("key", scientists[index]);
+      data.add(element);
+    }
+    return data;
+  }
+
+  @Test
```

Comment:
Isn't it what I do line 221 (batchSize is 5 and window contains 6 elements, so I test that it outputs 1 element)? Or do you mean testing that an element that comes between `window.maxTimestamp` and `window.maxTimestamp + allowedLateness` is correctly output in the batch?